### DISCUSSION TITLE
Questions about sidekiq batches

### DISCUSSION THREAD

**Discussion by rohithighspot** [sidekiq/sidekiq#5261] [Q&A] [answered]
Hi,
We are using Sidekiq with batches in a Kubernetes cluster and its been working great. Have some questions related to Sidekiq batches, would appreciate any information about this: 

1) Some of our batches have a lot of jobs (in the millions) where each job completes within milliseconds to a few seconds. We do need to get the on_success callback to know when the batch succeeded. Due to the high number of jobs, we don't add all the jobs in one batch.jobs block, but batch.jobs executes inside a loop and we add multiple jobs (in the hundreds) in each loop iteration. Ideally we want to set the on_success callback after all the jobs we need are added to the batch, i.e.; after we exit the loop that adds jobs to the batch. However, at this point the batch is immutable and we are not able to set the callback. If we set the callback before getting into the loop, we have seen at times that the on_success gets called even before we exit the loop. So the question is how/when should we set the batch callbacks such that we're sure that no more jobs will be added after the callback is set (for us this is the point when we exit the loop). 

2) Haven't been able to consistently reproduce this issue, but sometimes we see that a batch is stalled, with 1 or 2 pending jobs. The queue is empty and workers are idle but these jobs never show up in the busy tab on the web dashboard. Investigating in Redis we find that the set called "b-{batch-id}-jids has the pending jobs. A list with a name like "queue:sq|{worker-name}|{queue-name} has an item that represents the job we enqueued. However, this worker is idle and not executing this job that it has in its private queue. 
If we restart the workers for this queue, the job moves from the worker's private queue to the shared queue, and then the job gets picked up by newly started worker and completes successfully. 

Would appreciate any help about these 2 issues. 

thanks,
Rohit

**Comment by mperham** [comment, marked as answer]
1. You cannot call `batch.jobs` inside a loop. You may call `batch.jobs` **once** when starting the batch. Every further call for `batch.jobs` must happen inside a job within that batch. https://github.com/mperham/sidekiq/wiki/Batches#huge-batches

2. Yes, this is super_fetch rescuing orphaned jobs. These most often happen because of a poorly implemented deployment shutdown.  There are past issues which discuss k8s and shutdown, please search the issues. https://github.com/mperham/sidekiq/wiki/Deployment#overview
